### ISSUE TITLE
`groupBy()` doesn't accept it's own `GroupByArgs` type as an argument

### ISSUE THREAD

**Issue by lukadriel7** [prisma/prisma#7183]
### Bug description

After generating the client, the groupBy method of the client's object throw an error when using a parameter typed GroupByArgs 
i.e.
```
userGroup: UserGroupByArgs = {
    by: ["name"]
}
prismaClient.user.groupBy(userGroup);

```
will throw an error

```
src/users/users.resolver.ts:37:5 - error TS2615: Type of property 'AND' circularly references itself in mapped type '{ [K in keyof { AND?: UserScalarWhereWithAggregatesInput[]; OR?: UserScalarWhereWithAggregatesInput[]; NOT?: UserScalarWhereWithAggregatesInput[]; id?: IntWithAggregatesFilter; name?: StringWithAggregatesFilter; profile?: JsonWithAggregatesFilter; }]: Or<...> extends 1 ? { ...; }[K] extends infer TK ? GetHavingField...'.

37     this.prismaService.user.groupBy(group);
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/users/users.resolver.ts:37:5 - error TS2615: Type of property 'NOT' circularly references itself in mapped type '{ [K in keyof { AND?: UserScalarWhereWithAggregatesInput[]; OR?: UserScalarWhereWithAggregatesInput[]; NOT?: UserScalarWhereWithAggregatesInput[]; id?: IntWithAggregatesFilter; name?: StringWithAggregatesFilter; profile?: JsonWithAggregatesFilter; }]: Or<...> extends 1 ? { ...; }[K] extends infer TK ? GetHavingField...'.

37     this.prismaService.user.groupBy(group);
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/users/users.resolver.ts:37:5 - error TS2615: Type of property 'OR' circularly references itself in mapped type '{ [K in keyof { AND?: UserScalarWhereWithAggregatesInput[]; OR?: UserScalarWhereWithAggregatesInput[]; NOT?: UserScalarWhereWithAggregatesInput[]; id?: IntWithAggregatesFilter; name?: StringWithAggregatesFilter; profile?: JsonWithAggregatesFilter; }]: Or<...> extends 1 ? { ...; }[K] extends infer TK ? GetHavingField...'.

37     this.prismaService.user.groupBy(group);
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/users/users.resolver.ts:37:37 - error TS2345: Argument of type 'UserGroupByArgs' is not assignable to parameter of type '{ where?: UserWhereInput; orderBy?: UserOrderByWithAggregationInput[]; by: UserScalarFieldEnum[]; having?: UserScalarWhereWithAggregatesInput; ... 6 more ...; _max?: UserMaxAggregateInput; } & { ...; } & (`Error: Field "${any}" used in "having" needs to be provided in "by".` | [...])'.
  Type 'UserGroupByArgs' is not assignable to type '{ where?: UserWhereInput; orderBy?: UserOrderByWithAggregationInput[]; by: UserScalarFieldEnum[]; having?: UserScalarWhereWithAggregatesInput; ... 6 more ...; _max?: UserMaxAggregateInput; } & { ...; } & [...]'.
    Type 'UserGroupByArgs' is not assignable to type '{ orderBy: Enumerable<UserOrderByWithAggregationInput>; }'.
      Property 'orderBy' is optional in type 'UserGroupByArgs' but required in type '{ orderBy: Enumerable<UserOrderByWithAggregationInput>; }'.

37     this.prismaService.user.groupBy(group);
                                       ~~~~~

[00:22:13] Found 4 errors. Watching for file changes.
```

### How to reproduce

1. Create a project with prisma as a dependency
2. initialise the project and configure db connection
3. write schema and generate client
4. use groupBy function with parameters as a typed GroupByArgs
5. see error

### Expected behavior

No type error should be raised when using groupBy with parameters as a typed GroupByArgs

### Prisma information

Prisma schema
```
// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

datasource db {
  provider = "postgres"
  url      = env("DATABASE_URL")
}

generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["selectRelationCount", "orderByRelation", "filterJson"]
}

generator nestgraphql {
  provider   = "node node_modules/prisma-nestjs-graphql"
  output     = "../src/graphql"
  emitSingle = true
}

model User {
  id      Int    @id @default(autoincrement())
  name    String
  profile Json
}
```
Prisma client query
```
@Mutation(() => User)
  async createUser(
    @Args() params: CreateOneUserArgs,
    @Info() info: GraphQLResolveInfo,
  ) {
    const a: UserGroupByArgs = {
      by: [UserScalarFieldEnum.id],
    };
    const b: Prisma.UserGroupByArgs = {
      by: [Prisma.UserScalarFieldEnum.id],
    };
    this.prismaService.user.groupBy(b);
    const select = new PrismaSelect(info).value;
    params = { ...params, ...select };
    return this.prismaService.user.create(params);
  }
```

### Environment & setup

- OS: Lubuntu 20.04
- Database: Postgresql 13.
- Node.js version: 14.17.0

### Prisma Version

```
prisma               : 2.23.0
@prisma/client       : 2.23.0
Current platform     : debian-openssl-1.1.x
Query Engine         : query-engine adf5e8cba3daf12d456d911d72b6e9418681b28b (at node_modules/@prisma/engines/query-engine-debian-openssl-1.1.x)
Migration Engine     : migration-engine-cli adf5e8cba3daf12d456d911d72b6e9418681b28b (at node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Introspection Engine : introspection-core adf5e8cba3daf12d456d911d72b6e9418681b28b (at node_modules/@prisma/engines/introspection-engine-debian-openssl-1.1.x)
Format Binary        : prisma-fmt adf5e8cba3daf12d456d911d72b6e9418681b28b (at node_modules/@prisma/engines/prisma-fmt-debian-openssl-1.1.x)
Default Engines Hash : adf5e8cba3daf12d456d911d72b6e9418681b28b
Studio               : 0.393.0
Preview Features     : selectRelationCount, orderByRelation, filterJson
```

**Comment by pantharshit00** [comment]
Related to https://github.com/prisma/prisma/issues/6494

**Comment by matthewmueller** [comment]
I can confirm this issue:

```prisma
// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

model User {
  id   Int    @id
  name String
}
```

```ts
import { PrismaClient, Prisma } from "@prisma/client"
const prisma = new PrismaClient()

async function main() {
  await prisma.$connect()
  const args: Prisma.UserGroupByArgs = {
    by: ["name"],
  }
  prisma.user.groupBy(args)
}
```

Produces the error on `prisma.user.groupBy(args)`

```
index.ts:9:3 - error TS2615: Type of property 'AND' circularly references itself in mapped type '{ [K in keyof { AND?: Enumerable<UserScalarWhereWithAggregatesInput>; OR?: Enumerable<UserScalarWhereWithAggregatesInput>; NOT?: Enumerable<...>; id?: number | IntWithAggregatesFilter; name?: string | StringWithAggregatesFilter; }]: Or<...> extends 1 ? { ...; }[K] extends infer TK ? GetHavingFields<...> : never : {}...'.

9   prisma.user.groupBy(args)
```

```

```

**Comment by rs-payres** [comment]
We are not using groupBy but we came across this error in some tests when trying to mock the Prisma client.
`const mockPrismaClient = mockDeep<PrismaClient>();`
This has worked fine on <= 2.19, but we are now trying to upgrade past that and it caused type errors for our entire schema on this mock. I think this is because `groupBy` was made GA in the 2.20 release so it is now included in our Prisma client.

This code was written to test custom middleware, is there a better way we should be mocking the Prisma client for this type of testing?

**Comment by 13shubham** [comment]
Hi @rs-payres I am getting the same error while mocking the prisma for my test, do you happen to know the solution for that one ?
![image](https://user-images.githubusercontent.com/60516897/133281266-5d9b0b5b-0f0d-48e6-a001-eaf510ea30bd.png)

**Comment by Chiji1108** [comment]
Is there any update on this issue?

**Comment by omar-dulaimi** [comment]
Hello,

I'm facing this issue with Nestjs+Prisma stack. Any workarounds?

Or should I just stop using GroupBy and not expose it?

**Comment by jasonmacdonald** [comment]
I'm currently blocked by this. It seems like a pretty core piece of functionality that is completely broken.

**Comment by CaioCamatta** [comment]
I am also experiencing similar issues with `groupBy`, i.e. `Type of property 'OR'/'AND'/'NOT' circularly references...`. Running TypeScript 4.5.4 and  Node 16.13.0.

**Comment by Abdulrahmanelheyb** [comment]
I am also having same the error
![image](https://user-images.githubusercontent.com/45883300/168754828-1b82069c-e654-4885-8e37-1844301da7db.png)

**Comment by Mark-Panda** [comment]
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
//@ts-ignore
return await prisma.user.groupBy(args);


temporary plan

**Comment by johnkm516** [comment]
Is there an update on this issue? Ignoring errors might work when using only the Prisma Client types but there are code generators that use the DMMF for GraphQL resolvers and the current implementation of GroupBy makes it impossible to code-generate GroupBy queries. GraphQL type validations will throw an error.

**Comment by Mingyang-Li** [comment]
I'm still facing this issue, any update on this?

I can only use a temporary fix for it like below
```ts
public async groupBy<T extends Prisma.ExpenseGroupByArgs>(
    args: Prisma.SelectSubset<T, Prisma.ExpenseGroupByArgs>,
) {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  //@ts-ignore
  return await this.prisma.expense.groupBy(args);
}
```

**Comment by jameskennethrobinson** [comment]
Also facing this issue

**Comment by ludralph** [comment]
Closing this in favour of [17297](https://github.com/prisma/prisma/issues/17297)
